### PR TITLE
Info when publishing <3.0.0 constraint that is interpreted as <4.0.0

### PR DESCRIPTION
**Pull request by sigurdm**
Fix of https://github.com/dart-lang/pub/issues/3552

### CODE REVIEW

**Review by szakarias — APPROVED** [review]

File: lib/src/validator/sdk_constraint.dart [hunk 1]
```diff
@@ -58,6 +46,21 @@ class SdkConstraintValidator extends Validator {
...
+      if (
+          // We only want to give this hint if there was no other problems with
+          // the sdk constraint.
+          warnings.isEmpty &&
+              errors.isEmpty &&
+              originalConstraint != effectiveConstraint) {
+        hints.add('''
+The declared sdk constraint is '$originalConstraint', this is interpreted as '$effectiveConstraint'.
+
+Consider updating the sdk constraint to:
```
Comment (jonasfj) [inline lib/src/validator/sdk_constraint.dart:58]: Do we normally capitalize SDK..

```suggestion
The declared SDK constraint is '$originalConstraint', this is interpreted as '$effectiveConstraint'.

Consider updating the SDK constraint to:
```

File: lib/src/validator/sdk_constraint.dart [hunk 2]
```diff
@@ -58,6 +46,21 @@ class SdkConstraintValidator extends Validator {
             'See https://dart.dev/tools/pub/publishing#publishing-prereleases '
             'For more information on pre-releases.');
       }
+      if (
+          // We only want to give this hint if there was no other problems with
+          // the sdk constraint.
+          warnings.isEmpty &&
+              errors.isEmpty &&
+              originalConstraint != effectiveConstraint) {
```
Comment (jonasfj) [inline lib/src/validator/sdk_constraint.dart:54]: This will only happen in Dart 3.x, right? and this change won't make it into 2.x?

**Review by jonasfj — APPROVED** [review]

File: lib/src/validator/sdk_constraint.dart (hunk 2 above)
Comment (jonasfj) [reply lib/src/validator/sdk_constraint.dart:54]: Context: I don't think we want people to write `<4.0.0` before they are using a 3.0.0-... SDK.

Comment (sigurdm) [reply lib/src/validator/sdk_constraint.dart:54]: Yes, unless we cherry-pick it to the 2.19-series. The sdk main branch is already in dart 3-land.

File: lib/src/validator/sdk_constraint.dart (hunk 1 above)
Comment (sigurdm) [reply lib/src/validator/sdk_constraint.dart:58]: I guess so.

**Pull request merged**